README.md
simpleSAMLphp-casserver
=========================

# Usage

## Install

Install with composer

    composer require simplesamlphp/simplesamlphp-module-casserver

## Configuration

See the `config-templates` folder for examples of configuring this module

## Debug

To aid in debugging you can print out the CAS ticket xml rather then returning a ticket id.
Enable `debugMode` in `module_casserver.php` and then add a query parameter `debugMode=true` to the CAS login url.

Logging in to https://cas.example.com/cas/login?debugMode=true&service=http://localhost/ would now print the xml for that service.

```xml
<?xml version="1.0">
<cas:serviceResponse xmlns:cas="http://www.yale.edu/tp/cas">
 <cas:authenticationSuccess>
  <cas:user>testuser@example.com</cas:user>
  <cas:attributes>
   <cas:eduPersonPrincipalName>testuser@example.com</cas:eduPersonPrincipalName>
   <cas:base64Attributes>false</cas:base64Attributes>
  </cas:attributes>
 </cas:authenticationSuccess>
</cas:serviceResponse>
```

# Development

Run `phpcs` to check code style

    phpcs --standard=PSR2 lib/ tests/ www/ templates/

Run `phpunit` to test

    ./vendor/bin/phpunit
    
Use docker php image to easily test between versions

    docker run -ti --rm -v "$PWD":/usr/src/myapp  -w /usr/src/myapp php:7.1-cli ./vendor/bin/phpunit


You can auto correct some findings from phpcs. It is recommended you do this after stage your changes (or maybe even commit) since there is a non-trivial chance it will just mess up your code.

    phpcbf --ignore=somefile.php --standard=PSR2 lib/ tests/ www/ templates/
    
# History
CAS 1.0 and 2.0 compliant CAS server module for simpleSAMLphp

This is the simpleSAMLphp CAS server module developed at the State and University Library in Aarhus Denmark.
The module is a fork of an old version of the CAS module shipped with simpleSAMLphp which has undergone a couple of
iterations of refactoring, bugfixes and enhancements. For details see the ChangeLog in the doc directory.

All files are rewritten based on work by Dubravko Voncina. See Google Groups discussion in this thread:
http://groups.google.com/group/simplesamlphp/browse_thread/thread/4c655d169532650a

# License
This work is licensed under a Creative Commons GNU Lesser General Public License License.



docs/ChangeLog
/*
 *    simpleSAMLphp-casserver is a CAS 1.0 and 2.0 compliant CAS server in the form of a simpleSAMLphp module
 *
 *    Copyright (C) 2013  Bjorn R. Jensen
 *
 *    This library is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 2.1 of the License, or (at your option) any later version.
 *
 *    This library is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public
 *    License along with this library; if not, write to the Free Software
 *    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

Unreleased
    * Allow certain authproc filters to be used in configuration
    * Support method parameter at login
    * Improve handling of invalid xml element names for attributes
    * Minimum supported simplesamlphp version bumped to 1.17
    * debugMode option to display cas ticket xml
    * Allow per service overriding of configuration options for www/login

2018-07-20 Bjorn Rohde Jensen
    * Release 6.1.0
    * Minimum supported simplesamlphp version bumped to 1.15.
    * Fixed some deprecation warnings by using namespaces to refer to the relevant classes.
    * Added support for using Redis as a ticket store.

2016-08-01 Bjorn Rohde Jensen
    * Release 6.0.0
    * Renamed module from sbcasserver to casserver.
    * Added composer file to make module installable according to simplesamlphp guidelines.
    * Replaced use of deprecated SimpleSAMLphp api's.
    * Fixed a bug when comparing service url parameter to ticket service url, which caused service urls with
        %20 encoded spaces to fail verification.

2014-10-15 Bjørn Rohde Jensen <brj@statsbiblioteket.dk>
    * Release 5.2.3
    * Added a missing linefeed character in the CAS 1.0 ticket validation failure response.

2014-03-17 Bjørn Rohde Jensen <brj@statsbiblioteket.dk>
    * Release 5.2.2
    * Fixed a bug where the sanitized forms of url parameters were used for more than validation.
    * Fixed a bug in the sanitizer logic to remove ';jsessionid' from urls.

2014-02-27 Bjørn Rohde Jensen <brj@statsbiblioteket.dk>
    * Release 5.2.1
    * Fixed a bug where attributes containing colons in their names caused serviceValidate to return invalid xml.

2014-02-25 Bjørn Rohde Jensen <brj@statsbiblioteket.dk>
    * Release 5.2.0
    * Removed hardcoded removal of attributes starting with urn:oid.
    * Added a logged in landing page in case the client does not provide a service url.
    * It is now possible to add an attribute indicating whether the attributes are base64 encoded or not.

2014-02-07 Bjørn Rohde Jensen <brj@statsbiblioteket.dk>
    * Release 5.1.0
    * Fixed a bug where the 'url' parameter was required even when displaying a logged out page.
    * Fixed a bug where attribute transfer could not be disabled.
    * Added support for specifying which attributes to transfer.
    * Added support for specifying a named subset of IdPs allowing clients to restrict the list of IdPs in the
        wayf step.

2013-11-04 Bjørn Rohde Jensen <brj@statsbiblioteket.dk>
    * Release 5.0.0
    * Removed support for storing tickets in proprietary key-value store AttributeStore.
    * Added support for storing tickets in a SQL database.
    * Added support for storing tickets in memcached.
    * Added support for proxy tickets.
    * Added support for renewing login.
    * Added an optional logged out landing page with a return url.
    * Added support for specifying an idp during login thus skipping the wayf step.
    * Added support for specifying a language hint during login.

2013-01-29 Bjørn Rohde Jensen <brj@statsbiblioteket.dk>
    * Release 4.0.0
    * Moved all filters to a dedicated project.
    * Moved theme related code into the sbthemes project.
    * Added an abstract cas ticket store with concrete subclasses to store tickets locally or in the attribute store.
    * Changed cas 2.0 response generation from hand coded to using php dom document.
    * Fixed a bug in cas 1.0 response generation, where yes/no was returned in upper case.

2012-10-31 Bjørn Rohde Jensen <brj@statsbiblioteket.dk>
    * release 3.3.0
    * IPRoleMapper can now be configured to add its roles to any given attribute
        by setting the filter parameter 'targetAttributeName'.

2012-10-24 Bjørn Rohde Jensen <brj@statsbiblioteket.dk>
    * release 3.2
    * Added AttributeCollector auth filter for collecting attributes from
        the attribute store.
    * Added CAS logout

2012-05-22  Per Møldrup-Dalum  <pdj@statsbiblioteket.dk>
    * release 3.1
    * Corrected a spelling bug

2012-05-08  Per Møldrup-Dalum  <pdj@statsbiblioteket.dk>
    * Release 3
    * Added the Statsbiblioteket theme from sbdisco
    * Added the IP role mapper from sbdisco

2011-10-31  Per Møldrup-Dalum  <pdj@statsbiblioteket.dk>
    * Release 2.3
    * Added a hack to keep compatibility with the eduPersonNIN attribute.

2011-03-29  Per Møldrup-Dalum  <pdj@statsbiblioteket.dk>
    * Release 2.2
    * Reintroduced ignore urn:oid attributes.
    * Added configurable base 64 encoding of attribute values.
    * Removed debug messages.
    * Changed the casserver to use the configured auth source.

2011-01-18  Per Møldrup-Dalum  <pdj@statsbiblioteket.dk>
    * Changed default from disabled to enabled

2011-01-17  Per Møldrup-Dalum  <pdj@statsbiblioteket.dk>
    * Creating release 2.0
    * All files are rewritten based on work by Dubravko Voncina. See Google
        Groups discussion in this thread:
        http://groups.google.com/group/simplesamlphp/browse_thread/thread/4c655d169532650a
    * Creating release 2.1

www/login.php
<?php

/*
 *    simpleSAMLphp-casserver is a CAS 1.0 and 2.0 compliant CAS server in the form of a simpleSAMLphp module
 *
 *    Copyright (C) 2013  Bjorn R. Jensen
 *
 *    This library is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 2.1 of the License, or (at your option) any later version.
 *
 *    This library is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 *
 *    You should have received a copy of the GNU Lesser General Public
 *    License along with this library; if not, write to the Free Software
 *    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Incoming parameters:
 *  service
 *  renew
 *  gateway
 *  entityId
 *  scope
 *  language
 */

use SimpleSAML\Module\casserver\Cas\AttributeExtractor;
use SimpleSAML\Module\casserver\Cas\Ticket\TicketFactory;
use SimpleSAML\Module\casserver\Cas\Ticket\TicketStore;
use SimpleSAML\Configuration;
use SimpleSAML\Locale\Language;
use SimpleSAML\Logger;
use SimpleSAML\Module;
use SimpleSAML\Session;
use SimpleSAML\Utils\HTTP;

require_once('utility/urlUtils.php');

$forceAuthn = isset($_GET['renew']) && $_GET['renew'];
$isPassive = isset($_GET['gateway']) && $_GET['gateway'];
// Determine if client wants us to post or redirect the response. Default is redirect.
$redirect = !(isset($_GET['method']) && 'POST' === $_GET['method']);

$casconfig = Configuration::getConfig('module_casserver.php');

$legal_service_urls = $casconfig->getValue('legal_service_urls', []);

if (isset($_GET['service']) && !checkServiceURL(sanitize($_GET['service']), $legal_service_urls)) {
    $message = 'Service parameter provided to CAS server is not listed as a legal service: [service] = '.
        var_export($_GET['service'], true);
    Logger::debug('casserver:'.$message);

    throw new \Exception($message);
}

$as = new \SimpleSAML\Auth\Simple($casconfig->getValue('authsource'));

if (array_key_exists('scope', $_GET) && is_string($_GET['scope'])) {
    $scopes = $casconfig->getValue('scopes', []);

    if (array_key_exists($_GET['scope'], $scopes)) {
        $idpList = $scopes[$_GET['scope']];
    } else {
        $message = 'Scope parameter provided to CAS server is not listed as legal scope: [scope] = '.
            var_export($_GET['scope'], true);
        Logger::debug('casserver:'.$message);

        throw new \Exception($message);
    }
}

if (array_key_exists('language', $_GET) && is_string($_GET['language'])) {
    Language::setLanguageCookie($_GET['language']);
}

$ticketStoreConfig = $casconfig->getValue('ticketstore', ['class' => 'casserver:FileSystemTicketStore']);
$ticketStoreClass = Module::resolveClass($ticketStoreConfig['class'], 'Cas_Ticket');
/** @var $ticketStore TicketStore */
/** @psalm-suppress InvalidStringClass */
$ticketStore = new $ticketStoreClass($casconfig);

$ticketFactoryClass = Module::resolveClass('casserver:TicketFactory', 'Cas_Ticket');
/** @var $ticketFactory TicketFactory */
/** @psalm-suppress InvalidStringClass */
$ticketFactory = new $ticketFactoryClass($casconfig);

$session = Session::getSessionFromRequest();

$sessionTicket = $ticketStore->getTicket($session->getSessionId());
$sessionRenewId = $sessionTicket ? $sessionTicket['renewId'] : null;
$requestRenewId = isset($_REQUEST['renewId']) ? $_REQUEST['renewId'] : null;

if (!$as->isAuthenticated() || ($forceAuthn && $sessionRenewId != $requestRenewId)) {
    $query = [];

    if ($sessionRenewId && $forceAuthn) {
        $query['renewId'] = $sessionRenewId;
    }

    if (isset($_REQUEST['service'])) {
        $query['service'] = $_REQUEST['service'];
    }

    if (isset($_REQUEST['method'])) {
        $query['method'] = $_REQUEST['method'];
    }

    if (isset($_REQUEST['renew'])) {
        $query['renew'] = $_REQUEST['renew'];
    }

    if (isset($_REQUEST['gateway'])) {
        $query['gateway'] = $_REQUEST['gateway'];
    }

    if (array_key_exists('language', $_GET)) {
        $query['language'] = is_string($_GET['language']) ? $_GET['language'] : null;
    }

    $returnUrl = HTTP::getSelfURLNoQuery().'?'.http_build_query($query);

    $params = [
        'ForceAuthn' => $forceAuthn,
        'isPassive' => $isPassive,
        'ReturnTo' => $returnUrl,
    ];

    if (isset($_GET['entityId'])) {
        $params['saml:idp'] = $_GET['entityId'];
    }

    if (isset($idpList)) {
        if (sizeof($idpList) > 1) {
            $params['saml:IDPList'] = $idpList;
        } else {
            $params['saml:idp'] = $idpList[0];
        }
    }

    $as->login($params);
}

$sessionExpiry = $as->getAuthData('Expire');

if (!is_array($sessionTicket) || $forceAuthn) {
    $sessionTicket = $ticketFactory->createSessionTicket($session->getSessionId(), $sessionExpiry);

    $ticketStore->addTicket($sessionTicket);
}

$parameters = [];

if (array_key_exists('language', $_GET)) {
    $oldLanguagePreferred = Language::getLanguageCookie();

    if (isset($oldLanguagePreferred)) {
        $parameters['language'] = $oldLanguagePreferred;
    } else {
        if (is_string($_GET['language'])) {
            $parameters['language'] = $_GET['language'];
        }
    }
}

if (isset($_GET['service'])) {
    $attributeExtractor = new AttributeExtractor();
    $mappedAttributes = $attributeExtractor->extractUserAndAttributes($as->getAttributes(), $casconfig);

    $serviceTicket = $ticketFactory->createServiceTicket([
        'service' => $_GET['service'],
        'forceAuthn' => $forceAuthn,
        'userName' => $mappedAttributes['user'],
        'attributes' => $mappedAttributes['attributes'],
        'proxies' => [],
        'sessionId' => $sessionTicket['id']
    ]);

    $ticketStore->addTicket($serviceTicket);

    $parameters['ticket'] = $serviceTicket['id'];

    if (isset($_GET['debugMode']) && $_GET['debugMode'] == 'true' && $casconfig->getBoolean('debugMode', false)) {
        $method = 'serviceValidate';
        // Fake some options for validateTicket
        $_GET['ticket'] = $serviceTicket['id'];
        // We want to capture the output from echo used in validateTicket
        ob_start();
        require_once 'utility/validateTicket.php';
        $casResponse = ob_get_contents();
        ob_end_clean();
        echo '<pre>' . htmlspecialchars($casResponse) . '</pre>';
    } elseif ($redirect) {
        HTTP::redirectTrustedURL(HTTP::addURLParameters($_GET['service'], $parameters));
    } else {
        HTTP::submitPOSTData($_GET['service'], $parameters);
    }
} else {
    HTTP::redirectTrustedURL(
        HTTP::addURLParameters(Module::getModuleURL('casserver/loggedIn.php'), $parameters)
    );
}
